### ISSUE TITLE
Bugs in new FM cores

### ISSUE THREAD

**Issue by LeonardoDemartino** [mamedev/mame#8042]
@aaronsgiles

The new BSD FM core for the YM2612 doesn't do stereo for channel 5 while on dac mode.
https://github.com/mamedev/mame/blob/7118633dfdc23bb5cbaa2e3866043ed651d5e618/src/devices/sound/ym2612.cpp#L216

"Stereo output capability means that any of the sounds, FM or DAC, may be directed to the left, the right, or both outputs."
https://www.smspower.org/maxim/Documents/YM2612

Code of the old core doing stereo on channel 5 regardless of the dac_enabled flag:
https://github.com/mamedev/mame/blob/2ea2a582f39390bb4b2210c0ccc4455b57000749/src/devices/sound/fm2612.cpp#L2293

---

And another thing noticed is that the SSG-EG envelopes seems to be a lot faster.
This was discovered by comparing SSG-EG envelopes against [Nuked-OPN2](https://github.com/nukeykt/Nuked-OPN2) core which is a cycle accurate emulator.

Best regards and thanks for all the hard work!

**Comment by aaronsgiles** [comment]
Thanks for the feedback. I have the stereo fix implemented in my local branch.

As for the SSG-EG envelopes, do you have some examples? How did you detect the difference?

**Comment by LeonardoDemartino** [comment]
Hi Aaron! Thanks for your fast response!
Users of my tracker detected the difference when we started to do some tests with your cores.

Here you can find an example:
https://deflemask.com/ssg_bug.vgm
https://deflemask.com/ssg_bug_new_mame.wav
https://deflemask.com/ssg_ok_nuked.wav

Tell me if this is ok or your need more details/examples.

**Comment by aaronsgiles** [comment]
That's perfect, thanks, found the bug. I was overwriting the invert state when restarting the attack on each repeat in this case.

I really appreciate the detailed and knowledgeable feedback. Do you have more unit tests like this to run against emulator cores? I am happy to hear of legitimate problems.

BTW, I am currently restructuring the cores to make them less MAME-dependent. Current WIP is [here](https://github.com/aaronsgiles/ymfm). I'm still thinking about ways to make the interface simpler, so that part could still change, but already the code can be built now 100% outside of MAME, and includes a simple VGM renderer as an example.

**Comment by LeonardoDemartino** [comment]
So the fix is simply to delete this line?:
https://github.com/mamedev/mame/blob/7118633dfdc23bb5cbaa2e3866043ed651d5e618/src/devices/sound/ymfm.cpp#L2272

Thanks for the link to that repo! The cores were, indeed, too much MAME-dependent and I had a very hard time extirpating all the mame devices logic and audio streamers.

**Comment by aaronsgiles** [comment]
Ok, that one was more subtle, but important. Envelope rates were a little off for large rates. Should be fixed with the latest pushes.

**Comment by LeonardoDemartino** [comment]
Thanks a lot Aaron! <3

A new issue I faced a minute ago (sorry if I'm obsessed with this):
https://deflemask.com/ssg_4.vgm
https://deflemask.com/ssg_4_mame.wav
https://deflemask.com/ssg_4_nuked.wav

There is a click/noise at the end of the envelopes.
Here is an image of the issue: https://deflemask.com/mame_vs_nuked_4.jpg (up MAME, down NUKED)

**Comment by aaronsgiles** [comment]
I completed my investigation on this and it looks like I was right about it being a timing issue. I modified my vgmrender.cpp to support running nuked in parallel to compare, and also to support output at native sample rate, and this is what I get:
https://aarongiles.com/for/github/ssg_2_ymfm.wav
https://aarongiles.com/for/github/ssg_2_nuked.wav

They are extremely close and I don't see any clicks/noise in ymfm, unless I am missing something. My player is careful to only send one address/data pair per output sample to both cores, which is a crude approximation for how long the busy signal would be high.
